src/overlay/Peer.h

```diff
@@ -171,47 +165,84 @@ class Peer : public std::enable_shared_from_this<Peer>,
   protected:
     class MsgCapacityTracker : private NonMovableOrCopyable
     {
-        std::weak_ptr<Peer> mWeakPeer;
-        StellarMessage mMsg;
+        std::weak_ptr<Peer> const mWeakPeer;
+        StellarMessage const mMsg;
+        std::atomic<bool> mFinished{false};
 
       public:
         MsgCapacityTracker(std::weak_ptr<Peer> peer, StellarMessage const& msg);
-        ~MsgCapacityTracker();
         StellarMessage const& getMessage();
-        std::weak_ptr<Peer> getPeer();
+        // Since MsgCapacityTracker is passed around between thread and stored
+        // by ASIO in lambdas, do not rely on RAII to trigger completion.
+        // Instead, explicitly call `finish`.
+        void finish();
     };
 
     OverlayAppConnector mAppConnector;
 
     Hash const mNetworkID;
     std::shared_ptr<FlowControl> mFlowControl;
-    VirtualClock::time_point mLastRead;
-    VirtualClock::time_point mLastWrite;
-    VirtualClock::time_point mEnqueueTimeOfLastWrite;
+    std::atomic<VirtualClock::time_point> mLastRead;
+    std::atomic<VirtualClock::time_point> mLastWrite;
+    std::atomic<VirtualClock::time_point> mEnqueueTimeOfLastWrite;
 
     PeerRole const mRole;
     OverlayMetrics& mOverlayMetrics;
     PeerMetrics mPeerMetrics;
 
+    // Mutex to protect PeerState, which can be accessed and modified from
+    // multiple threads
+#ifndef USE_TRACY
+    std::recursive_mutex mutable mStateMutex;
+#else
+    mutable TracyLockable(std::recursive_mutex, mStateMutex);
+#endif
+
     HmacSha256Key mSendMacKey;
     HmacSha256Key mRecvMacKey;
-    uint64_t mSendMacSeq{0};
-    uint64_t mRecvMacSeq{0};
-    VirtualTimer mRecurringTimer;
-
+    std::atomic<uint64_t> mSendMacSeq{0};
+    std::atomic<uint64_t> mRecvMacSeq{0};
```

Comment:
I notice the mSendMacKey and mRecvMacKey aren't actually synchronized. Of course there's a sort of protocol-enforced boundary where they're only ever modified before we start running anything on a background thread, but that's an implicit contract it'd be easy to break with future refactoring. I wonder if it makes sense to try factoring-out the HMAC machinery into a helper class with its own lock, similar to the flow control class?